### PR TITLE
Aurora matmul

### PR DESCRIPTION
**Pull request by arifiorino**
Adds `MatMulBasicBlock` and unit tests.

### CODE REVIEW

**Comment by ddkang** [comment]
Do we have timing numbers?

**Comment by arifiorino** [comment]
Here are timing numbers (single threaded on the campus cluster):
| `log n` | `1xn` times `nxn` proof time (seconds) |
|--------|--------|
|3|0.209453201|
|4|0.289997053|
|5|0.438705079|
|6|0.68228832|
|7|1.067893454|
|8|1.825728348|
|9|3.275972258|
|10|6.175311896|

File: src/basic_block/matmul.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+use ark_bn254::{Bn254, Fr, G1Affine, G1Projective, G2Affine, G2Projective};
+use ark_ec::pairing::Pairing;
+use ark_ff::Field;
+use ark_poly::{univariate::DensePolynomial, EvaluationDomain, GeneralEvaluationDomain};
+use ark_std::{ops::Mul, ops::Sub, One, UniformRand, Zero};
+use ndarray::ArrayD;
+use rand::{rngs::StdRng, SeedableRng};
+
+// Inputs to basic block are v,r_0,r_1,... where r_0,r_1,... are the rows of a matrix M
+// Output of basic block is w = vM
```
Comment (Vimalakirti) [inline src/basic_block/matmul.rs:14]: I think the comment here should be w = Mv. Below is my reason:
Based on the script, given r = [alpha^0, alpha^1,...,alpha^n], 
Aurora checks rw=rMv, where rw=B and rMv=A.

File: src/basic_block/matmul.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+      Q_x: util::msm::<G1Projective>(&srs.0, &B_Q_poly.coeffs).into(),
+      zero_div: util::msm::<G1Projective>(&srs.0, &B_poly.coeffs[1..]).into(),
+    };
+
+    // Blinding
+    let mut rng2 = StdRng::from_entropy();
+    let r: Vec<_> = (0..7).map(|_| Fr::rand(&mut rng2)).collect();
+    let proof: Vec<G1Affine> = vec![A.x, A.Q_x, A.zero, A.zero_div, B.x, B.Q_x, B.zero_div];
+    let mut proof: Vec<G1Affine> = proof.iter().enumerate().map(|(i, x)| ((*x) + srs.0[srs.1.len() - 1] * r[i]).into()).collect();
+    let C = vec![
```
Comment (Vimalakirti) [inline src/basic_block/matmul.rs:110]: I think it would be helpful to add more comments to explain these blinding factors

File: src/tests.rs [hunk 3]
```diff
@@ -59,4 +59,14 @@ fn testBasicBlocks() {
...
+
+  let n: usize = 1 << 6;
+  let m: usize = 1 << 3;
+  let mut inputs: Vec<Vec<Fr>> = vec![];
+  for _ in 0..m + 1 {
+    inputs.push((0..n).into_par_iter().map_init(rand::thread_rng, |rng, _| Fr::rand(rng)).collect());
+  }
+  let inputs: Vec<ArrayD<Fr>> = inputs.iter().map(|x| arr1(&x).into_dyn()).collect();
+  let inputs: Vec<_> = inputs.iter().map(|x| x).collect();
+  testBasicBlock(MatMulBasicBlock {}, srs, &arr1(&vec![]).into_dyn(), &inputs);
```
Comment (Vimalakirti) [inline src/tests.rs:71]: I did run this test, everything works well. But how can I get the same bench statistics as you sent above? (where shall I add the timing script? Shall I bench Data::new?)

**Review by Vimalakirti — COMMENTED** [review]
Please check the comments

File: src/basic_block/matmul.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    // Calculate pow
+    let mut pow: Vec<Fr> = vec![Fr::one(); m];
+    for i in 0..m - 1 {
+      pow[i + 1] = pow[i] * alpha;
+    }
+    let pow_poly = DensePolynomial { coeffs: domain_m.ifft(&pow) };
+    let pow_x2 = util::msm::<G2Projective>(&srs.1, &pow_poly.coeffs);
+
+    // Calculate flat
+    let temp:Vec<_> = (0..m).map(|i|inputs[1 + i].g1).collect();
```
Comment (liliatangxy) [inline src/basic_block/matmul.rs:158]: rustfmt

**Review by liliatangxy — APPROVED** [review]

**Pull request closed without merging**